### PR TITLE
Cardano-db-sync module

### PR DESCRIPTION
**Pull request by zmrocze**
solves issue #58. 

`Cardano-db-sync` module. With `cardano-db-sync.postgres.enable` also runs postgres.
Single test that awaits sync progress to start.

### CODE REVIEW

**Comment by zmrocze** [comment]
eh, attic input fails to build with these nixpkgs. but still id say go ahead and review

**Comment by zmrocze** [comment]
our nixos tests run forever in CI. Locally it's maybe 25 secs for me. But previously the `db-sync` failed after reaching 100s timeout (which it reached after many minutes (?))

File: flake.nix [hunk 1]
```diff
@@ -12,12 +12,14 @@
       url = "github:input-output-hk/cardano-configurations/21249e0d5c68b4e8f3661b250aa8272a8785d678";
       flake = false;
     };
+    cardano-db-sync = {
+      url = "github:intersectmbo/cardano-db-sync?ref=13.2.0.1"; # compatible with 8.7.3
```
Comment (aciceri) [inline flake.nix:16]:
```suggestion
      url = "github:intersectmbo/cardano-db-sync/13.2.0.1"; # compatible with 8.7.3
```

You can also directly write this.

**Review by aciceri — COMMENTED** [review]
I cannot run the test locally using
```nix run .#apps.x86_64-linux.vmTests-cardano-db-sync```
I get the following error:
```
nix run .#apps.x86_64-linux.vmTests-cardano-db-sync
```
Check also #69 

Anyway the test seems good! :muscle: You said it works locally, right? Any idea why it doesn't work on CI? You increased the timeout and decreased the target percentage, both reasonable actions.

File: modules/cardano-db-sync.nix [hunk 2]
```diff
@@ -0,0 +1,188 @@
+{
+  config,
+  lib,
+  ...
+}:
+with lib; let
+  cfg = config.cardano.cardano-db-sync;
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:7]: `config.cardano.db-sync`

File: modules/cardano-db-sync.nix [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        description = "Path to the postgresql socket.";
+      };
+    };
+    _environment = mkOption {
+      default = config.services.cardano-node.environments.${config.cardano.network} or {hack = "<Cardano envirionment attribute set>";};
+      internal = true;
+      type = attrs;
+      description = "Attribute set describing the chosen network (mainnet, testnet, etc.). By default taken from `services.cardano-node.environments`.";
+    };
+    # `services.cardano-db-sync` module options:
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:60]: don't copy the options, use them directly

File: modules/cardano-db-sync.nix [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+          else 5432;
+        description = "Postgres database port. See also option socketDir `cardano.cardano-db-sync.database.socketdir`.";
+      };
+      socketdir = lib.mkOption {
+        type = lib.types.str;
+        default = "/run/postgresql";
+        description = "Path to the postgresql socket.";
+      };
+    };
+    _environment = mkOption {
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:54]: this option is not needed, use the value directly

File: modules/cardano-db-sync.nix [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            ProcSubset = "pid";
+            PrivateTmp = true;
+            ProtectControlGroups = true;
+            PrivateUsers = true;
+            LockPersonality = true;
+          };
+        };
+        assertions = [
+          {
+            assertion = config.cardano.node.enable;
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:160]: not necessarily, if the node is running on another machine

File: modules/cardano-db-sync.nix [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            message = "Cardano db sync requires `cardano.node.enable`.";
+          }
+          {
+            assertion = (! cfg.postgres.enable) || (cfg.database.name == cfg.database.user);
+            message = "When postgres is enabled, we use the ensureDBOwnership option which expects the user name to match db name.";
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:165]: do we?

File: modules/cardano-db-sync.nix [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+          enable = true;
+          # see assertions: this is same as user name
+          ensureDatabases = [cfg.database.name];
+          ensureUsers = [
+            {
+              name = "${cfg.database.name}";
+              ensureDBOwnership = true;
+            }
+          ];
+          authentication =
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:180]: use ensureDBOwnership instead

File: modules/default.nix [hunk 8]
```diff
@@ -43,6 +43,12 @@
         ./http.nix
       ];
     };
+    cardano-db-sync = {
```
Comment (brainrake) [inline modules/default.nix:46]: `db-sync`

File: tests/cardano-db-sync.nix [hunk 9]
```diff
@@ -0,0 +1,46 @@
+{
+  perSystem.vmTests.tests.cardano-db-sync = {
```
Comment (brainrake) [inline tests/cardano-db-sync.nix:2]: `db-sync`

File: modules/cardano-db-sync.nix [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+          port = 5432;
+          socketdir = "/run/postgresql";
+        };
+        ```
+      or enable the default postgresql service with `services.cardano-db-sync.postgres.enable` and possibly overwrite the `services.postgresql` options for your need.
+    '';
+    postgres = {
+      enable = mkEnableOption "Run postgres and connect dbsync to it.";
+    };
+    database = {
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:29]: don't copy the options, use them directly

**Review by brainrake — CHANGES_REQUESTED** [review]

File: modules/cardano-db-sync.nix (hunk 3 above)
Comment (brainrake) [reply modules/cardano-db-sync.nix:60]: only add an option when it does more than just set the same option in the service, eg. setting multiple options or starting new services

File: modules/cardano-db-sync.nix (hunk 5 above)
Comment (zmrocze) [reply modules/cardano-db-sync.nix:160]: Okey, then I'll also add socketPath as an option, as now there is: 

```nix
          inherit (config.cardano.node) socketPath;
```

File: modules/cardano-db-sync.nix [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        type = int;
+        default =
+          if cfg.postgres.enable
+          then config.services.postgresql.settings.port
+          else 5432;
+        description = "Postgres database port. See also option socketDir `cardano.cardano-db-sync.database.socketdir`.";
+      };
+      socketdir = lib.mkOption {
+        type = lib.types.str;
+        default = "/run/postgresql";
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply modules/cardano-db-sync.nix:50]: went with:

```
# if postgres is enabled, use first socket from postgresql settings or default to /run/postgresql
if cfg.postgres.enable then builtins.head ((config.services.postgresql.settings.unix_socket_directories or []) ++ [ "/run/postgresql" ]);
```

File: modules/cardano-db-sync.nix (hunk 3 above)
Comment (zmrocze) [reply modules/cardano-db-sync.nix:60]: wasnt sure about that myself, good

File: modules/cardano-db-sync.nix (hunk 6 above)
Comment (zmrocze) [reply modules/cardano-db-sync.nix:165]:
```nix
          ensureUsers = [
            {
              name = "${cfg.database.name}";
              ensureDBOwnership = true;
            }
          ];
```

this means that `cfg.database.name` has access to `cfg.database.name` database, so if `database.name` is different, then the user needs to overwrite these permissions himself. 
But I should make it a warning I now think.

File: modules/cardano-db-sync.nix (hunk 10 above)
Comment (zmrocze) [reply modules/cardano-db-sync.nix:29]: for `port` and `socketDir` we set better defaults so we repeat these 2 options. Maybe it's better then to repeat all of `database.<...>` options? for intuitive consistency

File: flake.nix [hunk 12]
```diff
@@ -12,12 +12,14 @@
...
+    };
 
     # Utilities
     attic = {
       url = "github:zhaofengli/attic";
-      inputs.nixpkgs.follows = "nixpkgs";
-      inputs.nixpkgs-stable.follows = "nixpkgs";
+      # doesn't build yet with our nixpkgs
     };
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply flake.nix:23]: ill fix at root https://github.com/hercules-ci/hercules-ci-effects/pull/165 if nixpkgs there allow

File: modules/cardano-db-sync.nix (hunk 10 above)
Comment (brainrake) [reply modules/cardano-db-sync.nix:36]: don't repeat the options, just set the defaults in the `cardano.db-sync` implementation

File: modules/cardano-db-sync.nix [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+          port = 5432;
+          socketdir = "/run/postgresql";
+        };
+        ```
+      or enable the default postgresql service with `services.cardano-db-sync.postgres.enable` and possibly overwrite the `services.postgresql` options for your need.
+    '';
+    postgres = {
+      enable = mkEnableOption "Run postgres and connect dbsync to it." // {default = true;};
+    };
+    nodeSocketPath = mkOption {
```
Comment (brainrake) [inline modules/cardano-db-sync.nix:29]: we don't need this option, set  `services.cardano-db-sync.socketPath =  cardano.node.socketPath` in the implementation

**Review by brainrake — CHANGES_REQUESTED** [review]

**Comment by zmrocze** [comment]
@brainrake i think it's fine now, would you check once more?
I simply removed the marked redundant options and the redundent postgres `authentication` setting (relic of some fix while experimenting). 

### Irrelevant note on the side:
Before I was under the impression that you're supposed to abstract away the underlying module (here `services.cardano-db-sync`) with your module (`cardano.db-sync`) - and make the user not touch the underlying module. I see it's differently with nixos modules, but it does make some sense, I mean it's more direct and doesn't repeat and in the end the module abstraction is always going to be leaky. On the downside the user needs to not only read the option docs but also the source including imports.

File: modules/cardano-db-sync.nix (hunk 6 above)
Comment (brainrake) [reply modules/cardano-db-sync.nix]: added assertion

**Comment by brainrake** [comment]
Thanks @zmrocze , that makes sense. I think you solved one part by mentoning `services.cardano-db-sync` in the `cardano.db.sync.enable` documentation. We can make it more explicit, or add a link. Another part is rendering the documentation for the `services.*` modules: https://github.com/mlabs-haskell/cardano.nix/pull/71 .

Let me explain. There were upstream modules for cardano-node and cardano-db-sync. Wrapping everything leads to a large amount of extra code with no apparent advantage except having all the options in the same namespace. So we went the opposite way, not wrapping anything, just adding the minimum options to support default use cases (sometimes just `enable`), and setting options to coordinate between modules. You can set up the default use cases with the very few options under `cardano.*`. If you need more configurability, you can use the modules.
Another advantage is that modules under `services.*` without any dependencies by just importing the file. They can be easily contributed upstream or to nixpkgs.

**Pull request merged**